src/Processors/QueryPlan/Optimizations/optimizeJoin.cpp

```diff
@@ -1058,6 +1063,128 @@ QueryPlan::Node chooseJoinOrder(QueryGraphBuilder query_graph_builder, QueryPlan
     return result;
 }
 
+/// Quick check if a node will definitely produce zero rows
+/// This is a lightweight check that avoids expensive estimateReadRowsCount
+/// It only checks for obvious cases without analyzing indexes
```

Comment:
We do `reading->getAnalyzedResult/reading->selectRangesToRead` above, so why do we claim "without analyzing indexes" ?